implement MonitorPortlet for pull-request-monitoring plugin

simonsymhoven opened the pull request:
Implement portlets for the https://github.com/jenkinsci/pull-request-monitoring-plugin.

Add one portlet for each analysis result which is added by the warnings-ng plugin to the actual run. Each portlet contains a sunburst diagram of fixed, outstanding and new issues. The new issues are clustered by severities, os it looks like this:

![Bildschirmfoto 2021-05-23 um 15 53 23](https://user-images.githubusercontent.com/26878018/119263749-99a92680-bbe0-11eb-9f1a-da5febbea7db.png)


- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your master branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

<!--
Put an `x` into the [ ] to show you have filled the information
-->

KalleOlaviNiemitalo commented on plugin/pom.xml:
```diff
@@ -231,6 +232,11 @@
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
+      <groupId>io.jenkins.plugins</groupId>
+      <artifactId>pull-request-monitoring</artifactId>
+      <version>${pull-request-monitoring.version}</version>
+    </dependency>
```
Could this dependency be optional? The more plugins server administrators have to install, the higher the risk that one of them has a vulnerability.

simonsymhoven replied:
Yes,  I think this can be marked as optional

simonsymhoven commented:
Implement portlets for the https://github.com/jenkinsci/pull-request-monitoring-plugin.

Add one portlet for each analysis result which is added by the warnings-ng plugin to the actual run. Each portlet contains a sunburst diagram of fixed, outstanding and new issues. The new issues are clustered by severities, os it looks like this:

![Bildschirmfoto 2021-05-18 um 14 33 56](https://user-images.githubusercontent.com/26878018/118651676-19e91980-b7e6-11eb-8c13-104e921542da.png)

KalleOlaviNiemitalo commented:
This seems redundant to me because Warnings NG already has graphs for the same information.
If I understand correctly, Pull Request Monitoring only adds graphs to builds of pull requests, while Warnings NG adds them to all builds.

Perhaps this will be more useful if users can also add other graphs to the dashboard. Currently though, this MonitorPortlet implementation for Warnings NG seems to be only one available. What other portlets have been planned?

simonsymhoven commented:
The idea of the plugin is to aggregate different delta metrics for pull requests and present them in a dashboard. Warning-ng is now the first plugin to provide such a portlet for this dashboard. that other developer can use the dashboard an integrate the result of the static analysis into their dashboards. The whole thing is being developed as part of a bachelor's thesis under the supervision of @uhafner. For more details I would like refer to the [Readme](https://github.com/jenkinsci/pull-request-monitoring-plugin/blob/master/README.md) or the [ux-sig](https://www.youtube.com/watch?v=F1ISpA7K0YA) meeting of 28 april.

KalleOlaviNiemitalo commented on plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:
```diff
@@ -0,0 +1,130 @@
+
+    @Override
+    public Optional<String> getDetailViewUrl() {
+        return Optional.of("../" + action.getUrlName());
```
I find the relative URL difficult to understand but it seems to work as intended.

This is called from [portlet.jelly](https://github.com/jenkinsci/pull-request-monitoring-plugin/blob/38077ebb55e65859a0e34183c4b0057764d0ea1d/src/main/resources/portlet/portlet.jelly#L25), which is referenced from [MonitoringDefaultAction/index.jelly](https://github.com/jenkinsci/pull-request-monitoring-plugin/blob/6e0e418a008f149fad739ec097bf8617ea3faff1/src/main/resources/io/jenkins/plugins/monitoring/MonitoringDefaultAction/index.jelly#L36). MonitoringDefaultAction is attached to a Run and [its getUrlName method](https://github.com/jenkinsci/pull-request-monitoring-plugin/blob/c1a90464474738e6b45d02c2272a631eae37c135/src/main/java/io/jenkins/plugins/monitoring/MonitoringDefaultAction.java#L57-L60) returns "pull-request-monitoring". So the ".." walks back to the Run and then you get to the ResultAction, which is also attached to the same Run.

Nothing to fix in this PR I think, but documenting the base URL in [MonitorPortlet.getDetailViewUrl](https://github.com/jenkinsci/pull-request-monitoring-plugin/blob/c1a90464474738e6b45d02c2272a631eae37c135/src/main/java/io/jenkins/plugins/monitoring/MonitorPortlet.java#L56-L63) might be useful, especially because the result of MonitorPortlet.getIconUrl is interpreted differently.

simonsymhoven replied:
Thats true. It is probably easier and more understandable if I change this to an absolute url so that would be `Optional.of(action.getAbsolutUrl())`. I will expand the documentation of the method, that there is no base url, do you agree?

KalleOlaviNiemitalo replied:
In the old wiki, [Hyperlinks in HTML](https://wiki.jenkins.io/display/JENKINS/Hyperlinks+in+HTML) recommends not using absolute URLs that include the scheme and host. Absolute paths are OK, though. I don't know whether the advice there is still valid.

simonsymhoven replied:
iconUrl and detailViewUrl using both the app relative url. As soon as I released it, I will update this PR. Thanks.

uhafner replied:
It is still valid to always use relative paths (and never use absolute paths). Otherwise using a reverse proxy will be quite difficult. What makes more sense here: let the pr-view plugin handle the navigation from the dashboard. A portlet should only define the portion of the path that starts after the given build URL. So here:

```suggestion
        return Optional.of(action.getUrlName());
```

KalleOlaviNiemitalo commented on plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:
```diff
@@ -0,0 +1,130 @@
+    @Override
+    @JavaScriptMethod
+    public String getId() {
+        return action.getResult().getId();
```
I wonder if this ID should also include "warnings", to reduce the risk of other plugins defining portlet IDs that clash with IDs that Warnings NG copies from AnalysisModelParserDescriptor. However, perhaps the URL would then clash anyway.

simonsymhoven replied:
Good point. In general it makes sense, if the artifact id of the plugin is added as a suffix.

uhafner commented on plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:
```diff
@@ -0,0 +1,130 @@
+     *          true if result has issues, else false.
+     */
+    public boolean hasIssues() {
+        return action.getResult().getFixedIssues().getSize() != 0
```
Since you are already extending the warning plugin: move that code to the `AnalysisResult` so the code here will just be `action.getResult().hasIssues()` (tell don't ask principle).

uhafner commented on plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:
```diff
@@ -0,0 +1,130 @@
+    /**
+     * The factory for the {@link PullRequestMonitoringPortlet}.
+     */
+    @Extension
```
If the dependency is marked as optional we need to mark the extension as optional as well
```suggestion
    @Extension(optional = true)
```

uhafner commented on plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:
```diff
@@ -0,0 +1,130 @@
+
+        @Override
+        public Collection<MonitorPortlet> getPortlets(final Run<?, ?> build) {
+            List<ResultAction> action = build.getActions(ResultAction.class);
```
plural: `actions`

uhafner commented on plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:
```diff
@@ -0,0 +1,130 @@
+
+        @Override
+        public String getDisplayName() {
+            return "Warnings Next Generation";
```
This should be part of `Messages.properties` in `io.jenkins.plugins.analysis.core.portlets`.

uhafner commented:
> Implement portlets for the https://github.com/jenkinsci/pull-request-monitoring-plugin.
> 
> Add one portlet for each analysis result which is added by the warnings-ng plugin to the actual run. Each portlet contains a sunburst diagram of fixed, outstanding and new issues. The new issues are clustered by severities, os it looks like this:
> 
> ![Bildschirmfoto 2021-05-18 um 14 33 56](https://user-images.githubusercontent.com/26878018/118651676-19e91980-b7e6-11eb-8c13-104e921542da.png)

Can you move that description to the PR description?

KalleOlaviNiemitalo commented on plugin/src/main/java/io/jenkins/plugins/analysis/core/model/AnalysisResult.java:
```diff
@@ -416,6 +416,19 @@ public Report getIssues() {
+     * Check if {@link AnalysisResult} has issues.
+     *
+     * @return
+     *          true if {@link AnalysisResult} has issues, else false.
```
Could explicitly say that this includes fixed issues.

KalleOlaviNiemitalo commented on plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:
```diff
@@ -0,0 +1,128 @@
+    @Override
+    @JavaScriptMethod
+    public String getId() {
+        return "warnings-ng-" + action.getResult().getId();
```
In <https://github.com/jenkinsci/pull-request-monitoring-plugin/commit/4273a8d58824b458b3f3cf02d80664e6a23ec0d2#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R249>, you wrote

> It is recommended to use the artifact id of the plugin or parts of it as suffix.

But this is instead a prefix. Perhaps the documentation is wrong.

simonsymhoven replied:
Thanks. Will fix it.

KalleOlaviNiemitalo commented on plugin/src/main/resources/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet/monitor.jelly:
```diff
@@ -0,0 +1,21 @@
+
+  <st:adjunct includes="io.jenkins.plugins.echarts"/>
+
+  <j:choose>
+    <j:when test="${!it.hasIssues()}">
+      <div id="${it.id}" data="${it.getResultIssuesJsonModel()}"
+           style="width: ${it.preferredWidth}px; height: ${it.preferredHeight}px;"/>
+    </j:when>
+    <j:otherwise>
+      <h1>${%zeroIssues.title}</h1>
+      <p>${%zeroIssues.description}</p>
+    </j:otherwise>
+  </j:choose>
```
The hasIssues check seems reversed to me, or maybe I just don't understand how j:when works.

simonsymhoven replied:
Ou of course, that's stupid..

uhafner reviewed (changes requested):
This is a very good starting point! 

I wonder if we could additionally add the quality gate result somewhere in the portly (if configured in the build).

simonsymhoven commented:
Tank for the review. I updated the description with the latest portlet and added the quality gate (only if its not NOT_BUILD and if the result has issues).

uhafner commented on plugin/src/main/java/io/jenkins/plugins/analysis/core/model/AnalysisResult.java:
```diff
@@ -417,16 +417,13 @@ public Report getIssues() {
-                || getNewIssues().getSize() != 0
-                || getOutstandingIssues().getSize() != 0;
+    public boolean isEmpty() {
+        return getTotals().getTotalSize() == 0;
```
```suggestion
        return getTotals().getTotalSize() + getTotals().getFixedSize() == 0;
```

uhafner commented on plugin/src/main/java/io/jenkins/plugins/analysis/core/portlets/PullRequestMonitoringPortlet.java:
```diff
@@ -96,13 +97,43 @@ public String getResultIssuesJsonModel() {
+     *          the image url of the icon.
+     */
+    public String getQualityGateResultIconUrl() {
+        return action.getResult().getQualityGateStatus().getResult().color.getImageOf("16x16");
```
This is still a bug in my plugin. I need to replace these icons with the new Jenkins SVG status images.

simonsymhoven replied:
So I can leave it like this for now?

uhafner replied:
Yes, nothing to do on your side. I just noticed that I need to fix it in the plugin (or in Jenkins core), I just forgot about it...

uhafner commented on plugin/src/test/java/io/jenkins/plugins/analysis/warnings/steps/StepsITest.java:
```diff
@@ -1009,6 +1012,31 @@ public void showPreventXxeSecurity656() {
+        JsonObject model = JsonParser.parseString(portlet.getResultIssuesAsJsonModel()).getAsJsonObject();
+        JsonObject newIssues = model.getAsJsonObject("new");
+
+        assertThat(model.get("fixed").getAsInt()).isZero();
```
Maybe using JSON assertions would be simpler:

```
assertThatJson(portlet.getResultIssuesAsJsonModel()).node("fixed").isEqualTo(0)
```

uhafner commented:
BTW: do not use force push after reviews. Otherwise the history in the comments is gone (and I need to recreate my local branches).

simonsymhoven commented:
> BTW: do not use force push after reviews. Otherwise the history in the comments is gone (and I need to recreate my local branches).

Sorry for that. I always using a rebase to squash the bunch of commits and update the pr before merging, it's a stupid habit..

uhafner commented:
> update the pr before merging

That still can be done using the other merge button.

simonsymhoven commented:
> > update the pr before merging
> 
> That still can be done using the other merge button.

Thats true.

uhafner reviewed (approved):
Thanks, looks good now!

Just some nitpicking: can you fix the 2 PMD warnings?

One thing we should discuss (maybe before we merge this PR) if it would be better if the extension point is an abstract class. That would make it easier to change the API in the future.

simonsymhoven commented:
> Thanks, looks good now!
> 
> 
> 
> Just some nitpicking: can you fix the 2 PMD warnings?
> 
> 
> 
> One thing we should discuss (maybe before we merge this PR) if it would be better if the extension point is an abstract class. That would make it easier to change the API in the future.

Oh, yes of course! 
Why should this make it easier if the factory is an abstract class?

uhafner commented:
> Why should this make it easier if the factory is an abstract class? 

Because interfaces are kind of sealed and cannot be changed after a release (otherwise all existing implementations will have compile errors after an upgrade). Abstract base classes can be extended more easily by adding new methods that contain a default implementation for existing consumers. Moreover you can put additional fields and methods in those base classes (with all access modifiers). Most extension points in Jenkins are defined in this way. (Or at least there is an interface **and** an abstract base class).

KalleOlaviNiemitalo commented:
Do interface default methods in Java 8 change that? I guess not entirely, because interfaces still cannot define instance fields nor private methods.

simonsymhoven commented:
> > Why should this make it easier if the factory is an abstract class?
> 
> Because interfaces are kind of sealed and cannot be changed after a release (otherwise all existing implementations will have compile errors after an upgrade). Abstract base classes can be extended more easily by adding new methods that contain a default implementation for existing consumers. Moreover you can put additional fields and methods in those base classes (with all access modifiers). Most extension points in Jenkins are defined in this way. (Or at least there is an interface **and** an abstract base class).

So you propose to wrap both interfaces into abstract classes and add default implementations there, which other plugins have to override? But then, I can not force the user to override the methods and do not use the default implementation unless every method return null?

uhafner commented:
> > > Why should this make it easier if the factory is an abstract class?
> > 
> > 
> > Because interfaces are kind of sealed and cannot be changed after a release (otherwise all existing implementations will have compile errors after an upgrade). Abstract base classes can be extended more easily by adding new methods that contain a default implementation for existing consumers. Moreover you can put additional fields and methods in those base classes (with all access modifiers). Most extension points in Jenkins are defined in this way. (Or at least there is an interface **and** an abstract base class).
> 
> So you propose to wrap both interfaces into abstract classes and add default implementations there, which other plugins have to override? But then, I can not force the user to override the methods and do not use the default implementation unless every method return null?

At least the portlet seems to be something that will change from time to time. Here an abstract class (and not an interface) makes more sense. You can define all methods without any meaningful default as `abstract` now. (And some methods like default size can return something useful). The point is just in *the future* new methods can be added without breaking something. When someone implements it the first time, those `abstract` methods are just fine. 

Existing examples in Jenkins: `IssuesTotalColumn extends ListViewColumn`, `IssuesChartPortlet extends DashboardPortlet`, etc. 

I'm not sure if the same applies to the `factory`, it currently contains only the method `Collection<MonitorPortlet> getPortlets(Run<?, ?> build);`, maybe we need to update this later with a method that accepts the PR as parameter, or the target branch, etc.? Honestly, I don't know if it will be stable. But it does not hurt much to convert the interfaces to an abstract class.

simonsymhoven commented:
> > > > Why should this make it easier if the factory is an abstract class?
> > > 
> > > 
> > > Because interfaces are kind of sealed and cannot be changed after a release (otherwise all existing implementations will have compile errors after an upgrade). Abstract base classes can be extended more easily by adding new methods that contain a default implementation for existing consumers. Moreover you can put additional fields and methods in those base classes (with all access modifiers). Most extension points in Jenkins are defined in this way. (Or at least there is an interface **and** an abstract base class).
> > 
> > 
> > So you propose to wrap both interfaces into abstract classes and add default implementations there, which other plugins have to override? But then, I can not force the user to override the methods and do not use the default implementation unless every method return null?
> 
> At least the portlet seems to be something that will change from time to time. Here an abstract class (and not an interface) makes more sense. You can define all methods without any meaningful default as `abstract` now. (And some methods like default size can return something useful). The point is just in _the future_ new methods can be added without breaking something. When someone implements it the first time, those `abstract` methods are just fine.
> 
> Existing examples in Jenkins: `IssuesTotalColumn extends ListViewColumn`, `IssuesChartPortlet extends DashboardPortlet`, etc.
> 
> I'm not sure if the same applies to the `factory`, it currently contains only the method `Collection<MonitorPortlet> getPortlets(Run<?, ?> build);`, maybe we need to update this later with a method that accepts the PR as parameter, or the target branch, etc.? Honestly, I don't know if it will be stable. But it does not hurt much to convert the interfaces to an abstract class.

I started to wrap the interfaces into abstract classes. Will come back here and update the pr after https://github.com/jenkinsci/pull-request-monitoring-plugin/pull/75 finished and released.
